pkvm-v6.18 fixes and improvements

maluka-dmytro opened the pull request:
- Make pkvm_hyp_donate_host() void and let it BUG_ON in case of any error, similarly to pkvm_host_unshare_hyp(), to facilitate catching bugs with incorrect usage of pkvm_hyp_donate_host() in pKVM code.
- Fix one such bug: when IPIv is disabled, pid_table is not allocated and not donated to pKVM, but we still try to undonate it when destroying the VM.

Commit:
SQUASHME: pKVM: x86: Make pkvm_hyp_donate_host() panic on any error

pKVM hypervisor should only call pkvm_hyp_donate_host() for those pages
that are guaranteed to be owned by the hypervisor (usually it means
pages that were previously donated from the host to the hypervisor via
pkvm_host_donate_hyp()), so if pkvm_hyp_donate_host() is called with
incorrect arguments, it means a bug in pKVM.

In practice we never check the return value of pkvm_hyp_donate_host(),
assuming that it should succeed.

So let pkvm_hyp_donate_host() BUG_ON in case of any error, and make it
void, similarly to pkvm_host_unshare_hyp().

Fixes: 40ad026472af ("pKVM: x86: Add memory page donation between host and hypervisor")
Signed-off-by: Dmytro Maluka <dmaluka@google.com>

Commit:
SQUASHME: pKVM: VMX: Don't undonate non-existent pid_table

If IPIv is disabled, pid_table is not allocated and not donated, so
it is incorrect to call pkvm_hyp_donate_host() to undonate it when
destroying the VM. So check if pid_table is not NULL before doing that.

Fixes: 36b120b99954 ("pKVM: VMX: Implement vcpu_precreate operation")
Signed-off-by: Dmytro Maluka <dmaluka@google.com>

cxdong commented on arch/x86/kvm/pkvm/mmu.c:
```diff
@@ -573,13 +568,11 @@ int pkvm_hyp_donate_host(unsigned long phys, unsigned long size, bool clear)
-	pkvm_host_mmu_unlock();
 
-	return ret;
+	pkvm_host_mmu_unlock();
```
It seems each function call in `pkvm_hyp_donate_host` should be wrapped by `BUG_ON()`. Does it look better if we just use `BUG_ON(ret)` here?

maluka-dmytro replied:
May be not a bad idea, yeah. Although it makes the code a bit longer, but OTOH it would make it easier for us to adjust this code if we decide to return this error in the future.

Pushed updated version with your suggestion. Also accordingly applied the same change to pkvm_host_unshare_hyp(), just to keep it consistent.

cxdong replied:
Thanks for the fixes. I will squash these patches.

Commit:
SQUASHME: pKVM: x86: Slightly refactor pkvm_host_unshare_hyp()

Cosmetic change: just to make pkvm_host_unshare_hyp() code consistent
with pkvm_host_donate_hyp(), BUG_ON on error when returning from the
function, not in every place where an error is encountered.

Fixes: 779ac62b6721 ("pKVM: x86: Add support to share/unshare host memory pages with pKVM")
Signed-off-by: Dmytro Maluka <dmaluka@google.com>
